### PR TITLE
Add APIv4 for `EntityFile`

### PR DESCRIPTION
**Pull request by dontub**
Overview
----------------------------------------
Adds APIv4 for `EntityFile`. https://lab.civicrm.org/dev/core/-/issues/2486#note_87390

Before
----------------------------------------
There was no APIv4 for `EntityFile`.

After
----------------------------------------
`EntityFile` can be accessed via APIv4.

### CODE REVIEW

**Comment by colemanw** [comment]
@civicrm-builder add to whitelist

File: xml/schema/Core/EntityFile.xml [hunk 1]
```diff
@@ -50,8 +50,14 @@
...
     <comment>FK to civicrm_file</comment>
     <html>
+      <type>Select</type>
       <label>File</label>
     </html>
+    <pseudoconstant>
+      <table>civicrm_file</table>
+      <keyColumn>id</keyColumn>
+      <labelColumn>id</labelColumn>
+    </pseudoconstant>
```
Comment (colemanw) [inline xml/schema/Core/EntityFile.xml:60]: I don't think we should add this psedoconstant. As a general rule, any field that is an FK usually does not also need a pseudoconstant as it's redundant and expensive to cache an entire (potentially very large) table in memory as an option list. Instead I think we should do:

```suggestion
      <type>EntityRef</type>
      <label>File</label>
    </html>
```

**Comment by colemanw** [comment]
Thanks @dontub I've left a couple review notes. I don't want to block this PR but I do think the schema in CiviCRM for storing files is a bit messed up and I don't love the idea of exposing that messed-upness to the API. In many cases the "entityFile" table is redundant with the file id already stored in the db (e.g. a custom field of type "file" already stores the file id) and I think the only really good use of this table is to 1-n relationships like file attachments for activities and notes.

**Comment by totten** [comment]
In the Gitlab discussion, Noah described `EntityFile` and `Attachment` as "light-weight" vs "heavy-weight". I think that's accurate. 

IIRC, in APIv3, the lightweight `EntityFile` was also done first - because, basically, it's easier. But personally I found using it fairly pernicious - because (a) it doesn't actually store the content; you have to find some other means to shuffle around the file content; (b) the permissions need to be driven by the referenced-record.

So there may be some limited use-cases where the lighter API is sufficient, e.g.

* Building a report for administrators to browse/audit files
* Implementing a server-side customization which attaches files

But generally, I think it's important to ensure APIv4's file management API addresses both content and the permissions.

File: xml/schema/Core/EntityFile.xml (hunk 1 above)
Comment (dontub) [reply xml/schema/Core/EntityFile.xml]: I've applied the suggestion. However this makes EntityFile unusable in APIv4 explorer. Choosing a file fails with `Entity File has no default label field.`

**Comment by dontub** [comment]
> I do think the schema in CiviCRM for storing files is a bit messed up and I don't love the idea of exposing that messed-upness to the API. In many cases the "entityFile" table is redundant with the file id already stored in the db (e.g. a custom field of type "file" already stores the file id) and I think the only really good use of this table is to 1-n relationships like file attachments for activities and notes.

Would you suggest not to use EntityFile to store 1-n (or m-n) relations in a custom extension?

**Comment by colemanw** [comment]
> Would you suggest not to use EntityFile to store 1-n (or m-n) relations in a custom extension?

That's probably fine. Have you tried the v3 Attachment API as a stopgap?

**Comment by dontub** [comment]
> > Would you suggest not to use EntityFile to store 1-n (or m-n) relations in a custom extension?
> 
> That's probably fine. Have you tried the v3 Attachment API as a stopgap?

As far as I can see the v3 Atttachment API allows only 1-1 relations.

**Comment by dontub** [comment]
@colemanw I've just rebased my branch. (One check fails because of a missing `since`.) Is there something I can do to further promote this?

**Comment by niels-heinemann** [comment]
Dear gurus, any news here? Is this something I can put some efforts in?

**Comment by colemanw** [comment]
@niels-heinemann @dontub we finally have an API for EntityFile thanks to #27379 - it is kind of like the v3 Attachment API but APIv4-flavored, with some in-sql calculated fields for `filename`, `url` `is_image` & `icon`.

**Comment by highfalutin** [comment]
That's awesome, @colemanw! Thank you.

**Pull request closed without merging**